### PR TITLE
Async fee accrue + async erc20 DAI to vat DAI

### PR DESCRIPTION
**Pull request by gbalabasquer**
Let me know what you think, this would solve one of the concerns Kurt had and would make each call to the flash mint module more gas efficient.

### CODE REVIEW

**Review by hexonaut — COMMENTED** [review]
So the code looks fine to me, but I'm not sure how I feel about this change. Personally I like avoiding adding another keeper cron job if possible, but I can see the appeal of gas-savings + getting the lost dai. I'm leaning more against this, but if you guys want to add it then I'm fine with that.

Care should be given for cases where `convert()` and `accrue()` are called inside the flash loan callback.

Maybe a middle ground is we add those functions to sweep up lost dai and put them behind a lock just to be extra safe?

**Comment by gbalabasquer** [comment]
Yes probably a `lock` modifier is the right thing to do, as they won't be gas sensitive anyway.
But yes, let's discuss if it is a change we like to have. To be honest this doesn't need to be called often at all.
Maybe once each 6 months or 1 year, so probably nothing to worry about the keepers perspective. It is something we could do ourselves or even include it in a spell once in a while.

File: src/flash.sol [hunk 1]
```diff
@@ -158,7 +161,7 @@ contract DssFlash is IERC3156FlashLender, IVatDaiFlashLender {
         dai.transferFrom(address(receiver), address(this), total);
         daiJoin.join(address(this), total);
         vat.heal(rad);
-        vat.move(address(this), vow, mul(fee, RAY));
+        require(vat.dai(address(this)) >= add(prev, fee), "DssFlash/ceiling-exceeded");
```
Comment (kmbarry1) [inline src/flash.sol]: I don't think the error message here is quite correct...

File: src/flash.sol [hunk 2]
```diff
@@ -183,8 +188,16 @@ contract DssFlash is IERC3156FlashLender, IVatDaiFlashLender {
         );
 
         vat.heal(amount);
-        vat.move(address(this), vow, fee);
+        require(vat.dai(address(this)) >= add(prev, fee), "DssFlash/ceiling-exceeded");
```
Comment (kmbarry1) [inline src/flash.sol]: ditto

File: src/flash.sol (hunk 1 above)
Comment (kmbarry1) [reply src/flash.sol]: I'm also not convinced this check is necessary. It wouldn't hurt if e.g. someone calls `accrue` during a flash loan.

**Comment by kmbarry1** [comment]
As I said in chat, I lean towards merging. This contract is already brutally heavy on storage operations, and this cuts out a cold write (to the Vow's `dai` balance). We can call `accrue` in the weekly spells.

**Review by kmbarry1 — APPROVED** [review]
LGTM modulo revert messages

File: src/flash.sol (hunk 1 above)
Comment (gbalabasquer) [reply src/flash.sol]: Yes the message is wrong and the check is necessary in this one but not in the other function as dai.transferFrom will fail if can't be done for the total.

Comment (gbalabasquer) [reply src/flash.sol]: To clarify the require is necessary on the vat flash loan as the only constraint is the `heal` which is only for the amount lent. However on the erc20 one the transferFrom will require the whole amount.

**Comment by gbalabasquer** [comment]
@kmbarry1 @hexonaut it should be ready for final review now

**Review by kmbarry1 — APPROVED** [review]
LGTM

**Review by hexonaut — APPROVED** [review]
LGTM. Let's merge this, and I'll put it up on Kovan.

**Pull request merged**